ngraph/core/reference/include/ngraph/runtime/reference/proposal.hpp

```diff
@@ -0,0 +1,453 @@
+                            const float scale_w =
+                                0.5f * (ratio_w * scales[scale] - coordinates_offset);
+                            const float scale_h =
+                                0.5f * (ratio_h * scales[scale] - coordinates_offset);
+
+                            // (x1, y1, x2, y2) for transformed box
+                            p_anchors_wm[scale] = center - scale_w;
+                            p_anchors_hm[scale] = center - scale_h;
+                            p_anchors_wp[scale] = center + scale_w;
+                            p_anchors_hp[scale] = center + scale_h;
+
+                            if (shift_anchors)
+                            {
+                                p_anchors_wm[scale] -= half_base_size;
+                                p_anchors_hm[scale] -= half_base_size;
+                                p_anchors_wp[scale] -= half_base_size;
+                                p_anchors_hp[scale] -= half_base_size;
+                            }
+                        }
+                    }
+                    return anchors;
+                }
+
+                template <typename T>
+                static void enumerate_proposals(const T* bottom4d,
+                                                const T* d_anchor4d,
+                                                const float* anchors,
+                                                std::vector<ProposalBox<T>>& proposals,
+                                                const int num_anchors,
+                                                const int bottom_H,
+                                                const int bottom_W,
+                                                const float img_H,
+                                                const float img_W,
+                                                const float min_box_H,
+                                                const float min_box_W,
+                                                const int feat_stride,
+                                                const float box_coordinate_scale,
+                                                const float box_size_scale,
+                                                float coordinates_offset,
+                                                bool initial_clip,
+                                                bool swap_xy,
+                                                bool clip_before_nms)
+                {
+                    const int bottom_area = bottom_H * bottom_W;
+
+                    const float* p_anchors_wm = anchors + 0 * num_anchors;
+                    const float* p_anchors_hm = anchors + 1 * num_anchors;
+                    const float* p_anchors_wp = anchors + 2 * num_anchors;
+                    const float* p_anchors_hp = anchors + 3 * num_anchors;
+
+                    for (size_t h = 0; h < bottom_H; ++h)
```

Comment:
Why do you use size_t here if `bottom_H` has a type int? What if I passed the negative value for bottom_H?

Reply:
bottom_H/W can't have negative values as they're deduced from input shape, which cannot have negative values.

Reply:
To better clarify in the code what I said earlier I will change their type to unsigned int.